File: app/src/main/java/com/divercity/app/core/base/UseCase.java
package com.divercity.app.core.base;

import io.reactivex.Observable;
import io.reactivex.Scheduler;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.observers.DisposableObserver;

public abstract class UseCase<T, Params> {

    private final CompositeDisposable compositeDisposable;
    private final Scheduler executorThread;
    private final Scheduler uiThread;

    public UseCase(Scheduler executorThread, Scheduler uiThread) {
        this.executorThread = executorThread;
        this.uiThread = uiThread;
        compositeDisposable = new CompositeDisposable();
    }

    public void execute(DisposableObserver<T> disposableObserver, Params params) {
        if (disposableObserver == null)
            throw new IllegalArgumentException("disposableObserver no debe ser null");

        final Observable<T> observable =
                this.createObservableUseCase(params).subscribeOn(executorThread).observeOn(uiThread);

        DisposableObserver observer = observable.subscribeWith(disposableObserver);
        compositeDisposable.add(observer);
    }

    public void dispose() {
        if (!compositeDisposable.isDisposed()) {
            compositeDisposable.dispose();
        }
    }

    protected abstract Observable<T> createObservableUseCase(Params params);

}

File: app/src/main/java/com/divercity/app/core/base/UseCase.kt
package com.divercity.app.core.base

import io.reactivex.Observable
import io.reactivex.Scheduler
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.observers.DisposableObserver

/**
 * Created by lucas on 01/03/2018.
 */
abstract class UseCase<T, Params>(private val executorThread: Scheduler, private val uiThread: Scheduler) {

    var compositeDisposable = CompositeDisposable()

    fun execute(disposableObserver: DisposableObserver<T>?, params: Params) {
       disposableObserver ?: throw IllegalArgumentException("disposableObserver cannot be null")

        val observable = this.createObservableUseCase(params).subscribeOn(executorThread).observeOn(uiThread)

        val observer = observable.subscribeWith(disposableObserver)
        compositeDisposable.add(observer)
    }

    fun dispose() {
        if (!compositeDisposable.isDisposed)
            compositeDisposable.dispose()
    }

    protected abstract fun createObservableUseCase(params : Params): Observable<T>
}